community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -19,109 +19,176 @@
-                        doSynchronizedWork();
-                    }
-                    finally
-                    {
-                        lock.unlock();
-                    }
-                }
+                return doSynchronizedWork( batch );
             }
-            catch ( InterruptedException e )
+            finally
             {
-                // We can't stop now, because our work has already been
-                // scheduled. So instead we're just going to reset the
-                // interruption status when we're done.
-                wasInterrupted = true;
+                unlock();
+                unparkAnyWaiters();
+                markAsDone( batch );
             }
         }
-        while ( !unit.done );
+        return null;
+    }
+
+    private void unparkAnyWaiters()
+    {
+        WorkUnit<Material,W> waiter = stack.get();
+        if ( waiter != stackEnd )
+        {
+            waiter.unpark();
+        }
+    }
+
+    private void checkFailure( Throwable failure ) throws ExecutionException
+    {
+        if ( failure != null )
+        {
+            throw new ExecutionException( failure );
+        }
+    }
+
+    private boolean tryLock( int tryCount, WorkUnit<Material,W> unit )
+    {
+        if ( lock.compareAndSet( null, Thread.currentThread() ) )
+        {
+            // Got the lock!
+            return true;
+        }
+
+        // Did not get the lock, spend some time until our work has either been completed,
+        // or we get the lock.
+        if ( tryCount < 10 )
+        {
+            // todo Java9: Thread.onSpinWait() ?
+            Thread.yield();
+        }
+        else
+        {
+            unit.park( 10, TimeUnit.MILLISECONDS );
+        }
+        return false;
+    }
 
-        if ( wasInterrupted )
+    private void unlock()
+    {
+        if ( lock.getAndSet( null ) != Thread.currentThread() )
         {
-            Thread.currentThread().interrupt();
+            throw new IllegalMonitorStateException(
+                    "WorkSync accidentally released a lock not owned by the current thread" );
         }
     }
 
-    private void doSynchronizedWork()
+    private WorkUnit<Material,W> grabBatch()
+    {
+        return reverse( stack.getAndSet( stackEnd ) );
+    }
+
+    private Throwable doSynchronizedWork( WorkUnit<Material,W> batch )
     {
-        WorkUnit<Material,W> batch = reverse( stack.getAndSet( stackEnd ) );
         W combinedWork = combine( batch );
+        Throwable failure = null;
 
         if ( combinedWork != null )
         {
-            combinedWork.apply( material );
+            try
+            {
+                combinedWork.apply( material );
+            }
+            catch ( Throwable throwable )
+            {
+                failure = throwable;
+            }
         }
-
-        markAsDone( batch );
+        return failure;
     }
 
     private WorkUnit<Material,W> reverse( WorkUnit<Material,W> batch )
```

Comment:
Given that work combining should be commutative, do we really need to reverse the stack of `WorkUnit`s before combining them?